src/connection.js

```diff
@@ -0,0 +1,112 @@
+const { Map, fromJS } = require('immutable')
+const OpSet = require('./op_set')
+
+// Returns true if all components of clock1 are less than or equal to those of clock2.
+// Returns false if the clocks are incomparable or if clock1 is greater than clock2.
+function lessOrEqual(clock1, clock2) {
+  return clock1.keySeq().concat(clock2.keySeq()).reduce(
+    (result, key) => (result && clock1.get(key, 0) <= clock2.get(key, 0)),
+    true)
+}
+
+function mergeClock(clockMap, docId, clock) {
+  clock = clockMap.get(docId, Map()).mergeWith((x, y) => Math.max(x, y), clock)
+  return clockMap.set(docId, clock)
+}
+
+// Keeps track of the communication with one particular peer. Allows updates for many documents to
+// be multiplexed over a single connection.
+//
+// To integrate a connection with a particular networking stack, two functions are used:
+// * `sendMsg` (callback passed to the constructor, will be called when local state is updated)
+//   takes a message as argument, and sends it out to the remote peer.
+// * `receiveMsg` (method on the connection object) should be called by the network stack when a
+//   message is received from the remote peer.
+//
+// The documents to be synced are managed by a `DocSet`. Whenever a document is changed locally,
+// call `setDoc()` on the docSet. The connection registers a callback on the docSet, and it figures
+// out whenever there are changes that need to be sent to the remote peer.
+//
+// knownClock is the most recent VClock that we think the peer has (either because they've told us
+// that it's their clock, or because it corresponds to a state we have sent to them on this
+// connection). Thus, everything more recent than knownClock should be sent to the peer.
+//
+// advertisedClock is the most recent VClock that we've advertised to the peer (i.e. where we've
+// told the peer that we have it).
+class Connection {
+  constructor (docSet, sendMsg) {
+    this._docSet = docSet
+    this._sendMsg = sendMsg
+    this._knownClock = Map()
+    this._advertisedClock = Map()
+    this._docChangedHandler = this.docChanged.bind(this)
+  }
+
+  open () {
+    for (let docId of this._docSet.docIds) this.docChanged(docId, this._docSet.getDoc(docId))
+    this._docSet.registerHandler(this._docChangedHandler)
+  }
+
+  close () {
+    this._docSet.unregisterHandler(this._docChangedHandler)
+  }
+
+  sendMsg (docId, clock, changes) {
+    const msg = {docId, clock: clock.toJS()}
+    this._advertisedClock = mergeClock(this._advertisedClock, docId, clock)
+    if (changes) msg.changes = changes.toJS()
+    this._sendMsg(msg)
+  }
+
+  maybeSendChanges (docId) {
+    const doc = this._docSet.getDoc(docId)
+    const clock = doc._state.getIn(['opSet', 'clock'])
+
+    if (this._knownClock.has(docId)) {
```

Comment:
We only send peers changes for documents which we know that they know about? Since the peer is about to send you a clock and ask for everything anyway, is this test useful?

Reply:
When we first establish a connection, we don't know whether the peer has any data at all for a given docId. If the peer already has all the data, or most of the data, we want to avoid sending them stuff they already have. So it's necessary to wait until we hear from the peer as to how much, if anything, they already have for some docId.